HealthVault/Classes/Caching/MHVThingCache.m

```diff
@@ -1,20 +1,410 @@
+        [self syncWithCompletionHandler:^(UIBackgroundFetchResult result) { }];
+    }
+    return self;
+}
+
+- (void)deauthorizedApplication
+{
+    [self.database deleteDatabase];
+
+    @synchronized (self.syncCompletionHandlers)
+    {
+        [self.syncCompletionHandlers removeAllObjects];
+    }
+}
+
+#pragma mark - Cached Results
+
+- (void)cachedResultsForQueries:(MHVThingQueryCollection *)queries
+                       recordId:(NSUUID *)recordId
+                     completion:(void(^)(MHVThingQueryResultCollection *_Nullable resultCollection))completion;
+{
+    id<MHVCachedRecord> record = [self.database fetchCachedRecord:recordId.UUIDString];
+    if (!record)
+    {
+        completion(nil);
+        return;
+    }
+
+    // If offline, dates don't apply & always use cached data
+    if (self.networkStatus.hasNetworkConnection)
+    {
+        // Check if it's been too long since the last sync
+        NSDate *lastSyncDate = [self.database lastSyncDateFromRecord:record];
+        if (fabs([lastSyncDate timeIntervalSinceNow]) > self.cacheConfiguration.maxCacheValidSeconds)
+        {
+            MHVLOG(@"Cache data is too old");
+            
+            //Start a sync so cache will get updated
+            [self syncWithCompletionHandler:^(UIBackgroundFetchResult result) { }];
+            
+            completion(nil);
+            return;
+        }
+    }
+    
+    //TODO...use cache
+    completion(nil);
+}
+
+#pragma mark - Syncing
+
+- (void)startSyncingForRecordId:(NSUUID *)recordId completion:(void (^_Nullable)(BOOL success))completion
+{
+    MHVASSERT_PARAMETER(recordId);
+    
+    if (!recordId)
+    {
+        if (completion)
+        {
+            completion(NO);
+        }
+        return;
+    }
+    
+    id<MHVCachedRecord> record = [self.database fetchCachedRecord:recordId.UUIDString];
+    if (!record)
+    {
+        MHVLOG(@"Record not found, adding new Sync record %@", [recordId.UUIDString substringToIndex:4]);
+        
+        record = [self.database newRecord:recordId.UUIDString];
+    
+        // Start syncing the new record; or if a sync is in progress it will wait until that completes before syncing the record
+        [self runBlockWhenNotSyncing:^(UIBackgroundFetchResult result)
+         {
+             [self syncRecords:@[record] completion:^(UIBackgroundFetchResult result)
+              {
+                  if (completion)
+                  {
+                      completion(result != UIBackgroundFetchResultFailed);
+                  }
+              }];
+         }];
+    }
+    else
+    {
+        if (completion)
+        {
+            completion(YES);
+        }
+    }
+}
+
+- (void)stopSyncingForRecordId:(NSUUID *)recordId
+{
+    MHVASSERT_PARAMETER(recordId);
+
+    if (!recordId)
+    {
+        return;
+    }
+    [self.database deleteRecord:recordId.UUIDString];
+}
+
+- (void)syncWithCompletionHandler:(void (^)(UIBackgroundFetchResult result))completionHandler
```

Comment:
Using UIBackgroundFetchResult is limiting your design here. I mentioned in the app delegate file, that there should be some abstraction between the cache and the performFetchWithCompletionHandler: method, and here is another example why. We need to provide more detail back to the caller in the form of errors or enums or something that will help debug issues, provide a way to test failures, and give the code a chance to react appropriately based on the problem. The fetchCachedRecords call to the database could have an error, but we are just logging it. That means that if there is an error we can't actually do anything to recover from it because we are just swallowing it.